Tests nmap replace

gperciva opened the pull request.

gperciva commented:
First draft; would allow #87 container-based Travis-CI.  (although if we want valgrind testing, that would negate this possibility as they haven't whitelisted valgrind: https://github.com/travis-ci/apt-package-whitelist/issues/3635)

Still, it would be nice not to have the extra dependency for the test suite.

cperciva commented on tests/spiped_nc/graceful_shutdown.h:
```diff
@@ -0,0 +1,18 @@
+
+/**
+ * graceful_shutdown_register(callback, caller_cookie):
+ * Initializes an event which checks for SIGTERM every 1.0
```
Why is this line wrapping so early?

cperciva commented on tests/spiped_nc/graceful_shutdown.c:
```diff
@@ -0,0 +1,96 @@
+#include "graceful_shutdown.h"
+
+struct graceful_shutdown_cookie {
+	void (* begin_shutdown)(void *);
```
I would have called these `callback` and `cookie` personally.  Not a big deal, but what does the verbosity get you?

cperciva replied:
Also, we usually have callbacks return `int`, so that they can report if something goes fatally wrong.

gperciva replied:
As for `cookie`... you'd like to see this?

struct cookie {
        int (* begin_shutdown)(void *);
        void * cookie;
};

Having two different `cookie` names seemed a bit... "off".

gperciva replied:
Also, looking at `graceful_shutdown.h`:

```
void * graceful_shutdown_register(callback, caller_cookie);
void graceful_shutdown_shutdown(cookie);
```

Wouldn't it be nice to distinguish between the "graceful_shutdown cookie", and "the cookie we pass to the shutdown callback" ?

cperciva commented on tests/spiped_nc/graceful_shutdown.c:
```diff
@@ -0,0 +1,96 @@
+
+	/* Use the callback function, or schedule another check in 1 second. */
+	if (should_shutdown)
+		G->begin_shutdown(G->caller_cookie);
```
Where is `G` freed?

gperciva replied:
`G` is freed in `graceful_shutdown_shutdown()`, which is called by `main.c` in the success and failure (`err4`) cases.

cperciva commented on tests/spiped_nc/graceful_shutdown.c:
```diff
@@ -0,0 +1,96 @@
+};
+
+static volatile sig_atomic_t should_shutdown = 0;
+static void * graceful_shutdown_timer_cookie = NULL;
```
Why have this here, but the cookie and callback as part of a structure which gets allocated on the heap, anyway?

cperciva commented on tests/spiped_nc/graceful_shutdown.c:
```diff
@@ -0,0 +1,96 @@
+	if (should_shutdown)
+		G->begin_shutdown(G->caller_cookie);
+	else
+		graceful_shutdown_timer_cookie = events_timer_register_double(
```
If this fails, we should return -1.

cperciva commented on tests/spiped_nc/graceful_shutdown.c:
```diff
@@ -0,0 +1,96 @@
+graceful_shutdown_shutdown(void * cookie)
+{
+
+	if (graceful_shutdown_timer_cookie != NULL)
```
Restoring the old signal handler would be a polite thing to do.  Of course, you would need to remember what it was.

cperciva commented on tests/spiped_nc/main.c:
```diff
@@ -0,0 +1,128 @@
+	FILE * out;
+};
+
+/* Forward definition. */
```
Why not static?

cperciva commented on tests/spiped_nc/main.c:
```diff
@@ -0,0 +1,128 @@
+
+	/* Parse command-line arguments. */
+	if (argc < 3) {
+		fprintf(stderr, "%s PORT FILENAME\n", argv[0]);
```
`usage:`

cperciva commented on tests/spiped_nc/simple_server.c:
```diff
@@ -0,0 +1,363 @@
+	struct conn_list_node * prev;
+	struct conn_list_node * next;
+	struct accept_state * A;
+	/* Reading a network message. */
```
I would separate the sections with a blank line her.e

gperciva commented:
Updated PR; changed everything you mentioned other than keeping `name_cookie` so we can avoid:

```
struct cookie {
        ...
        void * cookie;
};
```

If it looks good, I'll rewrite history.

cperciva commented on tests/spiped_nc/graceful_shutdown.c:
```diff
@@ -0,0 +1,122 @@
+ * detected, it calls the ${callback} function and passes it the ${cookie}.
+ */
+void *
+graceful_shutdown_register(int (* begin_shutdown)(void *),
```
Is it your intention that more than one graceful shutdown callback might be registered at once?

If yes, you need to make sure you restore signal handlers in the right order.

If no, why bother returning a cookie when you could just keep all the parameters as static variables inside this file?

gperciva commented:
Closing this until we get #104 added.  I'm now investigating the OSX Travis-CI flakiness in more detail.

gperciva commented:
This includes #110 and #104; see those before looking at this.  I'm just re-opening this PR so that we can keep track of this, and to schedule the Travis-CI build.

gperciva commented:
I'm sure that this won't get in tonight, but at least it's open now so it's easier to keep "in mind".

gperciva commented:
Updated this PR to remove code that you'd probably ask me to remove anyway.  (i.e. `simple_server` now no longer has `separate _init()`, `_run()`, `_shutdown()`; now it's just a single function).

cperciva commented on tests/nc-server/simple_server.h:
```diff
@@ -0,0 +1,16 @@
+ * ${caller_cookie}, along with the message.  Automatically shuts itself down
+ * after ${shutdown_after} connections have been dropped.
+ */
+int simple_server(const uint16_t, const size_t, const size_t,
```
`port` should be a `in_port_t`.  But why not simply pass the entire address string in?

cperciva replied:
BTW the consts are meaningless here.  These are pass-by-value so the function can't modify the original variables anyway.

gperciva replied:
Passing in the entire address string: I'd rather do the command-line parsing in the command-line parsing area.  If I passed in argv[1], then if they mistyped a port number, the error would come from `simple_server()`, not the beginning of `int main()`.

cperciva commented on tests/nc-server/simple_server.h:
```diff
@@ -0,0 +1,16 @@
+#include <stdint.h>
+
+/**
+ * simple_server_init(port, nconn_max, shutdown_after, callback, caller_cookie):
```
_init ?

cperciva commented on tests/nc-server/simple_server.c:
```diff
@@ -0,0 +1,368 @@
+ * simple_server_shutdown(cookie):
+ * Stop and free memory associated with the ${cookie}.
+ */
+void
```
This can now be static, right?

cperciva commented on tests/nc-server/main.c:
```diff
@@ -0,0 +1,106 @@
+int
+main(int argc, char ** argv)
+{
+	struct nc_cookie * C;
```
Why not simply make this a structure on the stack; what do you gain by making this a pointer and the allocating the structure?

cperciva commented on tests/nc-server/main.c:
```diff
@@ -0,0 +1,106 @@
+	uint16_t port;
+	const char * filename;
+
+	(void)argc; /* UNUSED */
```
Not unused.

cperciva commented on tests/nc-server/main.c:
```diff
@@ -0,0 +1,106 @@
+	}
+
+	/* Open the output file; can be /dev/null. */
+	if ((C->out = fopen(filename, "w+b")) == NULL) {
```
Why `w+b` rather than simply `wb`?  Are you expecting to read from this file?

cperciva commented on tests/nc-server/simple_server.c:
```diff
@@ -0,0 +1,368 @@
+
+	/* Schedule reading from this connection. */
+	if ((node_new->network_read_cookie = network_read(node_new->sock_read,
+	    node_new->buf, BUFLEN, 0, callback_read, node_new)) == NULL) {
```
You should ask for at least 1 byte.

cperciva commented on tests/nc-server/simple_server.c:
```diff
@@ -0,0 +1,368 @@
+	}
+
+	/* We have gained a connection. */
+	A->nconn += 1;
```
Maybe this should be rolled back in the error path if we fail to allocate a node or start reading from the socket?

cperciva commented on tests/nc-server/simple_server.c:
```diff
@@ -0,0 +1,368 @@
+		A->callback_nc_message(A->caller_cookie, R->buf,
+		    (size_t)lenread);
+
+		/* We can re-use R. */
```
`/* Try to read some more data. */` might be clearer?

gperciva commented:
Updated PR

cperciva commented on tests/nc-server/simple_server.c:
```diff
@@ -0,0 +1,372 @@
+err2:
+	free(node_new);
+err1:
+	A->nconn += 1;
```
Umm, what?

cperciva commented on tests/nc-server/simple_server.c:
```diff
@@ -0,0 +1,372 @@
+	return (-1);
+}
+
+/**
```
Static functions go before non-static functions unless there's a very good reason to do otherwise.

cperciva commented:
Looks reasonable.  Please tidy up the commit history.

gperciva commented:
Updated PR with rebased history

cperciva commented:
Looks good.  I'll merge when I notice that the tests have passed...
